README.md
This is an update of Eastwood, an offline java-based chart clone of Google Charts. It was discontinued in 2008. Westwood picks up where it left off.
========

A google image charts clone


eastwood/NEWS
Eastwood 1.1.0
--------------
18-Jul-2008

The second release of Eastwood Charts adds some important new features, 
some matching features from the Google Chart API and others extending it,
including:

- support for the sparkline chart type ('cht=ls');
- support for chart data scaling ('chds');
- new custom chart types supporting a 3D-effect on bar charts - just add a '3' 
      to the corresponding bar chart type ('cht=bvg3/bhg3/bvs3/bhs3');
- new 'ewd2' and 'ewtr' tags to support trendlines on bar charts and scatter
      plots;
- configurable fonts for the servlet;

The source code has been refactored to allow the chart engine to be reused 
outside of a servlet environment, for example in a client application (see the
org.jfree.eastwood.Application class) or an applet (see 
org.jfree.eastwood.ChartApplet).
  
In addition to the new features in this release, a number of bugs have been 
eliminated, and we've added some JUnit tests. 

The licence for the Eastwood has been changed to the GNU Lesser General Public 
Licence version 2.1 or later, matching the licence used for JFreeChart.  This
will mean (hopefully) we spend less time talking about licence terms, and more 
time coding.

Special thanks to Niklas Therning who has made several important contributions 
to this release.

Patches:
2003913 - disable plot border;
2002341 - 'chdl' and 'chdlp' support;
2001783 - support for UTF-8 and '%' in labels;
2001586 - Maven pom.xml.

Bug Fixes:
1872190 - pie chart label background paint;
1901369 - 'chdl' not working for bar charts;
1908089 - % not working for axis labels.


Eastwood 1.0.0
--------------
13 December 2007

Welcome to the first release of the Eastwood Chart Servlet.  This servlet 
emulates the recently released Google Chart API, using code that you install 
and run on your own server.  The advantage of this is that you don't HAVE to 
rely on Google to generate your charts.  The disadvantage is that you don't 
GET to rely on Google to generate your charts :)

The Eastwood Chart Servlet is Free Software (licensed under the terms of the
GNU Affero General Public Licence), so you have complete control over the 
chart generation process - you can modify it, fix bugs, tweak settings, or just
run it as is.

PLEASE NOTE THAT THIS INITIAL RELEASE ALMOST CERTAINLY CONTAINS BUGS.  We 
wanted to get this release out quickly (the Google Chart API was released just 
one week ago), so please report any issues you encounter.
    
Thanks!

Dave Gilbert
Eastwood Project Leader
http://www.jfree.org/eastwood/



eastwood/tests/org/jfree/eastwood/junit/ChartEngineTests.java
/* ===============
 * Eastwood Charts
 * ===============
 *
 * (C) Copyright 2007, 2008, by Object Refinery Limited.
 *
 * Project Info:  http://www.jfree.org/eastwood/index.html
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 2.1 of the License, or
 * (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
 * USA.
 *
 * [Java is a trademark or registered trademark of Sun Microsystems, Inc.
 * in the United States and other countries.]
 *
 * ---------------------
 * ChartEngineTests.java
 * ---------------------
 * (C) Copyright 2008, by Object Refinery Limited.
 *
 * Original Author:  David Gilbert (for Object Refinery Limited);
 *
 * Changes
 * -------
 * 27-Jun-2008 : Version 1 (DG);
 *
 */

package org.jfree.eastwood.junit;

import java.io.UnsupportedEncodingException;
import java.util.Map;

import junit.framework.Test;
import junit.framework.TestCase;
import junit.framework.TestSuite;

import org.jfree.eastwood.ChartEngine;
import org.jfree.eastwood.Parameters;

/**
 * Tests for the {@link ChartEngine} class.
 */
public class ChartEngineTests extends TestCase {

    /**
     * Returns the tests as a test suite.
     *
     * @return The test suite.
     */
    public static Test suite() {
        return new TestSuite(ChartEngineTests.class);
    }

    /**
     * Constructs a new set of tests.
     *
     * @param name  the name of the tests.
     */
    public ChartEngineTests(String name) {
        super(name);
    }

    /**
     * A test for bug report 2001830.
     *
     * @throws UnsupportedEncodingException
     */
    public void test2001830() throws UnsupportedEncodingException {
    	String paramStr = "cht=bhs&chco=ff0000|00ff00|0000ff,&chs=200x125" +
    	        "&chd=s:elo&chxt=x,y&chxl=1:|Dec|Nov|Oct|0:||20K||60K||100K|";
    	Map params = Parameters.parseQueryString(paramStr);
    	/* JFreeChart chart = */ ChartEngine.buildChart(params);
    }

}

eastwood/tests/org/jfree/eastwood/junit/DataUtilitiesTests.java
/* ===============
 * Eastwood Charts
 * ===============
 *
 * (C) Copyright 2007, 2008, by Object Refinery Limited.
 *
 * Project Info:  http://www.jfree.org/eastwood/index.html
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 2.1 of the License, or
 * (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
 * USA.
 *
 * [Java is a trademark or registered trademark of Sun Microsystems, Inc.
 * in the United States and other countries.]
 *
 * ------------------------
 * GDataUtilitiesTests.java
 * ------------------------
 * (C) Copyright 2008, by Object Refinery Limited.
 *
 * Original Author:  David Gilbert (for Object Refinery Limited);
 *
 * Changes
 * -------
 * 14-Jul-2008 : Version 1 (DG);
 *
 */

package org.jfree.eastwood.junit;

import java.util.List;

import junit.framework.Test;
import junit.framework.TestCase;
import junit.framework.TestSuite;

import org.jfree.data.category.CategoryDataset;
import org.jfree.data.general.PieDataset;
import org.jfree.data.xy.XYDataset;
import org.jfree.eastwood.DataUtilities;

/**
 * Tests for the {@link DataUtilities} class.
 */
public class DataUtilitiesTests extends TestCase {

    /**
     * Returns the tests as a test suite.
     *
     * @return The test suite.
     */
    public static Test suite() {
        return new TestSuite(DataUtilitiesTests.class);
    }

    /**
     * Constructs a new set of tests.
     *
     * @param name  the name of the tests.
     */
    public DataUtilitiesTests(String name) {
        super(name);
    }

    /**
     * A test to confirm that the parsing of data in the simple format is
     * working.
     */
    public void testParseSimpleData() {
    	String s = "s:ATb19,Mn5t_z";
    	List data = DataUtilities.parseSimpleData(s);
    	assertEquals(2, data.size());
    	List s1 = (List) data.get(0);
    	assertEquals(5, s1.size());
        assertEquals(0.0, ((Double) s1.get(0)).doubleValue(), EPSILON);
        assertEquals(19.0 / 61.0, ((Double) s1.get(1)).doubleValue(), EPSILON);
        assertEquals(27.0 / 61.0, ((Double) s1.get(2)).doubleValue(), EPSILON);
        assertEquals(53.0 / 61.0, ((Double) s1.get(3)).doubleValue(), EPSILON);
        assertEquals(1.0, ((Double) s1.get(4)).doubleValue(), EPSILON);
    	List s2 = (List) data.get(1);
    	assertEquals(6, s2.size());
        assertEquals(12.0 / 61.0, ((Double) s2.get(0)).doubleValue(), EPSILON);
        assertEquals(39.0 / 61.0, ((Double) s2.get(1)).doubleValue(), EPSILON);
        assertEquals(57.0 / 61.0, ((Double) s2.get(2)).doubleValue(), EPSILON);
        assertEquals(45.0 / 61.0, ((Double) s2.get(3)).doubleValue(), EPSILON);
        assertNull(s2.get(4));
        assertEquals(51.0 / 61.0, ((Double) s2.get(5)).doubleValue(), EPSILON);
    }

    /**
     * A test to confirm that the parsing of data in the extended format is
     * working.
     */
    public void testParseExtendedData() {
    	String s = "e:.Z__Aa";
    	List data = DataUtilities.parseExtendedData(s);
    	assertEquals(1, data.size());
    	List s1 = (List) data.get(0);
    	assertEquals(3, s1.size());
        assertEquals(4057.0 / 4095.0, ((Double) s1.get(0)).doubleValue(),
        		EPSILON);
        assertNull(s1.get(1));
        assertEquals(26.0 / 4095.0, ((Double) s1.get(2)).doubleValue(),
        		EPSILON);
    }

    /**
     * Some checks for the parseTextData() method.
     */
    public void testParseTextData() {
    	String s = "t:10.0,58.0,95.0|30.0,8.0,63.0";
    	List data = DataUtilities.parseTextData(s, new float[] {0.0f, 100.0f});
    	assertEquals(2, data.size());
    	List s1 = (List) data.get(0);
    	assertEquals(3, s1.size());
    	assertEquals(0.10, ((Float) s1.get(0)).doubleValue(), EPSILON);
    	assertEquals(0.58, ((Float) s1.get(1)).doubleValue(), EPSILON);
    	assertEquals(0.95, ((Float) s1.get(2)).doubleValue(), EPSILON);
    	List s2 = (List) data.get(1);
    	assertEquals(3, s2.size());
    	assertEquals(0.30, ((Float) s2.get(0)).doubleValue(), EPSILON);
    	assertEquals(0.08, ((Float) s2.get(1)).doubleValue(), EPSILON);
    	assertEquals(0.63, ((Float) s2.get(2)).doubleValue(), EPSILON);
    }

    /**
     * The Google data spec says to specify a missing value as '-1', but
     * trial and error suggests that any negative value is treated as a
     * missing value...and any number greater than 100.0 is capped at
     * 100.
     */
    public void testParseTextData2() {
    	String s = "t:-1.0,50.0,100.0,1000.0,-1000.0";
    	List data = DataUtilities.parseTextData(s, new float[] {0.0f, 100.0f});
    	assertEquals(1, data.size());
    	List s1 = (List) data.get(0);
    	assertEquals(5, s1.size());
    	assertNull(s1.get(0));
    	assertEquals(0.50, ((Float) s1.get(1)).doubleValue(), EPSILON);
    	assertEquals(1.00, ((Float) s1.get(2)).doubleValue(), EPSILON);
    	assertEquals(1.00, ((Float) s1.get(2)).doubleValue(), EPSILON);
    	assertNull(s1.get(0));
    }

    /**
     * Some checks for the parsePieDataset() method.
     */
    public void testParsePieDataset() {
    	PieDataset d = DataUtilities.parsePieDataset("s:Hello_World", "");
    	assertEquals(11, d.getItemCount());
    	assertEquals(7.0 / 61.0, d.getValue(0).doubleValue(), EPSILON);
    	assertEquals(30.0 / 61.0, d.getValue(1).doubleValue(), EPSILON);
    	assertEquals(37.0 / 61.0, d.getValue(2).doubleValue(), EPSILON);
    	assertEquals(37.0 / 61.0, d.getValue(3).doubleValue(), EPSILON);
    	assertEquals(40.0 / 61.0, d.getValue(4).doubleValue(), EPSILON);
    	assertNull(d.getValue(5));
    	assertEquals(22.0 / 61.0, d.getValue(6).doubleValue(), EPSILON);
    	assertEquals(40.0 / 61.0, d.getValue(7).doubleValue(), EPSILON);
    	assertEquals(43.0 / 61.0, d.getValue(8).doubleValue(), EPSILON);
    	assertEquals(37.0 / 61.0, d.getValue(9).doubleValue(), EPSILON);
    	assertEquals(29.0 / 61.0, d.getValue(10).doubleValue(), EPSILON);
    }

    /**
     * Some basic checks for the parseCategoryDataset() method.
     */
    public void testParseCategoryDataset() {
    	CategoryDataset d = DataUtilities.parseCategoryDataset("s:Hello_World",
    			"");
    	assertEquals(1, d.getRowCount());
    	assertEquals(11, d.getColumnCount());
    	assertEquals(7.0 / 61.0, d.getValue(0, 0).doubleValue(), EPSILON);
    	assertEquals(30.0 / 61.0, d.getValue(0, 1).doubleValue(), EPSILON);
    	assertEquals(37.0 / 61.0, d.getValue(0, 2).doubleValue(), EPSILON);
    	assertEquals(37.0 / 61.0, d.getValue(0, 3).doubleValue(), EPSILON);
    	assertEquals(40.0 / 61.0, d.getValue(0, 4).doubleValue(), EPSILON);
    	assertNull(d.getValue(0, 5));
    	assertEquals(22.0 / 61.0, d.getValue(0, 6).doubleValue(), EPSILON);
    	assertEquals(40.0 / 61.0, d.getValue(0, 7).doubleValue(), EPSILON);
    	assertEquals(43.0 / 61.0, d.getValue(0, 8).doubleValue(), EPSILON);
    	assertEquals(37.0 / 61.0, d.getValue(0, 9).doubleValue(), EPSILON);
    	assertEquals(29.0 / 61.0, d.getValue(0, 10).doubleValue(), EPSILON);
    }

    /**
     * Some basic checks for the parseXYDataset() method.
     */
    public void testParseXYDataset() {
    	XYDataset d = DataUtilities.parseXYDataset("s:Hello_World", "");
    	assertEquals(1, d.getSeriesCount());
    	assertEquals(11, d.getItemCount(0));
    	assertEquals(0.0, d.getXValue(0, 0), EPSILON);
    	assertEquals(0.1, d.getXValue(0, 1), EPSILON);
    	assertEquals(0.2, d.getXValue(0, 2), EPSILON);
    	assertEquals(0.3, d.getXValue(0, 3), EPSILON);
    	assertEquals(0.4, d.getXValue(0, 4), EPSILON);
    	assertEquals(0.5, d.getXValue(0, 5), EPSILON);
    	assertEquals(0.6, d.getXValue(0, 6), EPSILON);
    	assertEquals(0.7, d.getXValue(0, 7), EPSILON);
    	assertEquals(0.8, d.getXValue(0, 8), EPSILON);
    	assertEquals(0.9, d.getXValue(0, 9), EPSILON);
    	assertEquals(1.0, d.getXValue(0, 10), EPSILON);

    	assertEquals(7.0 / 61.0, d.getYValue(0, 0), EPSILON);
    	assertEquals(30.0 / 61.0, d.getYValue(0, 1), EPSILON);
    	assertEquals(37.0 / 61.0, d.getYValue(0, 2), EPSILON);
    	assertEquals(37.0 / 61.0, d.getYValue(0, 3), EPSILON);
    	assertEquals(40.0 / 61.0, d.getYValue(0, 4), EPSILON);
    	assertNull(d.getY(0, 5));
    	assertEquals(22.0 / 61.0, d.getYValue(0, 6), EPSILON);
    	assertEquals(40.0 / 61.0, d.getYValue(0, 7), EPSILON);
    	assertEquals(43.0 / 61.0, d.getYValue(0, 8), EPSILON);
    	assertEquals(37.0 / 61.0, d.getYValue(0, 9), EPSILON);
    	assertEquals(29.0 / 61.0, d.getYValue(0, 10), EPSILON);
    }

    /**
     * Some basic checks for the parseXYDataset2() method.
     */
    public void testParseXYDataset2() {
    	XYDataset d = DataUtilities.parseXYDataset2("s:Hello,Wor_d", "");
    	assertEquals(1, d.getSeriesCount());
    	assertEquals(5, d.getItemCount(0));

    	assertEquals(7.0 / 61.0, d.getXValue(0, 0), EPSILON);
    	assertEquals(30.0 / 61.0, d.getXValue(0, 1), EPSILON);
    	assertEquals(37.0 / 61.0, d.getXValue(0, 2), EPSILON);
    	assertEquals(37.0 / 61.0, d.getXValue(0, 3), EPSILON);
    	assertEquals(40.0 / 61.0, d.getXValue(0, 4), EPSILON);
    	assertEquals(22.0 / 61.0, d.getYValue(0, 0), EPSILON);
    	assertEquals(40.0 / 61.0, d.getYValue(0, 1), EPSILON);
    	assertEquals(43.0 / 61.0, d.getYValue(0, 2), EPSILON);
    	assertNull(d.getY(0, 3));
    	assertEquals(29.0 / 61.0, d.getYValue(0, 4), EPSILON);
    }

    private static final double EPSILON = 0.0000001;

}

eastwood/source/org/jfree/eastwood/DataUtilities.java
/* ===============
 * Eastwood Charts
 * ===============
 *
 * (C) Copyright 2007, 2008, by Object Refinery Limited.
 *
 * Project Info:  http://www.jfree.org/eastwood/index.html
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 2.1 of the License, or
 * (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
 * USA.
 *
 * [Java is a trademark or registered trademark of Sun Microsystems, Inc.
 * in the United States and other countries.]
 *
 * ------------------
 * DataUtilities.java
 * ------------------
 * (C) Copyright 2008, by Object Refinery Limited.
 *
 * Original Author:  David Gilbert (for Object Refinery Limited);
 * Contributors:     Niklas Therning (patch 2001599);
 *
 * Changes
 * -------
 * 14-Jul-2008 : Version 1, methods factored out of ChartEngine.java (DG);
 * 14-Jul-2008 : Added data scaling support - see patch 2001599 by
 *               Niklas Therning (DG);
 *
 */

package org.jfree.eastwood;

import java.text.CharacterIterator;
import java.text.StringCharacterIterator;
import java.util.Iterator;
import java.util.List;

import org.jfree.data.category.DefaultCategoryDataset;
import org.jfree.data.general.DefaultPieDataset;
import org.jfree.data.general.PieDataset;
import org.jfree.data.xy.XYDataset;
import org.jfree.data.xy.XYSeries;
import org.jfree.data.xy.XYSeriesCollection;

/**
 * A utility class for reading data that uses the Google chart encoding.
 */
public class DataUtilities {

    /**
     * Parses a string that uses the "simple" data encoding from the Google
     * Chart API.  The method returns a list of lists, with the sublists
     * each containing the data values for one series, the values being
     * represented by Double objects in the range 0.0 to 1.0.  Missing values
     * (which are encoded as '_') are returned as null entries in the list.
     *
     * @param s  the data string (<code>null</code> not permitted).
     *
     * @return A list of sublists.
     */
    public static List parseSimpleData(String s) {
    	if (s == null) {
    		throw new IllegalArgumentException("Null 's' argument.");
    	}
        List dataset = new java.util.ArrayList();
        StringCharacterIterator iterator = new StringCharacterIterator(s, 2);
        List series = new java.util.ArrayList();
        while (iterator.current() != CharacterIterator.DONE) {
            char c = iterator.current();
            if (c == ',') {
                dataset.add(series);
                series = new java.util.ArrayList();
            }
            else {
                int v = googleValue(c);
                if (v < Integer.MAX_VALUE) {
                    series.add(new Double(v / 61.0));
                }
                else {
                    series.add(null);
                }
            }
            iterator.next();
        }
        dataset.add(series);
        return dataset;
    }

    /**
     * Parses a string that uses the "extended" data encoding from the Google
     * Chart API.  The method returns a list of lists, with the sublists
     * each containing the data values for one series, the values being
     * represented by Double objects in the range 0.0 to 1.0.  Missing values
     * (which are encoded as '__') are returned as null entries in the list.
     *
     * @param s  the data string (<code>null</code> not permitted).
     *
     * @return A list of sublists.
     */
    public static List parseExtendedData(String s) {
    	if (s == null) {
    		throw new IllegalArgumentException("Null 's' argument.");
    	}
        List dataset = new java.util.ArrayList();
        StringCharacterIterator iterator = new StringCharacterIterator(s, 2);
        List series = new java.util.ArrayList();
        while (iterator.current() != CharacterIterator.DONE) {
            char c1 = iterator.current();
            if (c1 == ',') {
                dataset.add(series);
                series = new java.util.ArrayList();
            }
            else {
                char c2 = iterator.next();
                int c1v = googleValueExt(c1);
                int c2v = googleValueExt(c2);
                if (c1v < Integer.MAX_VALUE && c2v < Integer.MAX_VALUE) {
                    series.add(new Double((c1v * 64.0 + c2v) / 4095.0));
                }
                else {
                    series.add(null);
                }
            }
            iterator.next();
        }
        dataset.add(series);
        return dataset;
    }

    /**
     * Parses a string with text data encoding.  The method returns a list of
     * lists, with the sublists each containing the data values for one series,
     * the values being represented by Float objects in the range 0.0 to 1.0.
     * Missing values (which are encoded as any values outside the range 0.0 to
     * 100.0) are returned as null entries in the list.
     *
     * @param dataStr  the data string.
     * @param scaling  the scaling parameters.
     *
     * @return A list containing a sublist of Float objects for each series.
     */
    public static List parseTextData(String dataStr, float[] scaling) {
        List dataset = new java.util.ArrayList();
        StringCharacterIterator iterator = new StringCharacterIterator(dataStr,
                2);
        List series = new java.util.ArrayList();
        int baseIndex = 2;
        float min = getScalingMin(scaling, 0);
        float max = getScalingMax(scaling, 0);
        while (iterator.current() != CharacterIterator.DONE) {
            char c = iterator.current();
            if (c >= '0' && c <= '9' || c == '-' || c == '.') {
                iterator.next();
            }
            else if (c == ',') {
            	float val = (Float.parseFloat(dataStr.substring(baseIndex,
            			iterator.getIndex())) - min)/ (max - min);
            	if (val >= 0.0f) {
                    series.add(new Float(Math.min(val, 1.0f)));
            	}
            	else {
            		series.add(null);
            	}
                baseIndex = iterator.getIndex() + 1;
                iterator.next();
            }
            else if (c == '|') {
                series.add(new Float((Float.parseFloat(dataStr.substring(
                        baseIndex, iterator.getIndex())) - min) / (max - min)));
                baseIndex = iterator.getIndex() + 1;
                dataset.add(series);
                series = new java.util.ArrayList();
                iterator.next();
            	min = getScalingMin(scaling, dataset.size());
            	max = getScalingMax(scaling, dataset.size());
            }
            else {
                throw new IllegalArgumentException(
                        "'data' contains bad characters.");
            }
        }
        series.add(new Float((Float.parseFloat(dataStr.substring(
                baseIndex, iterator.getIndex())) - min) / (max - min)));
        dataset.add(series);
        return dataset;
    }

    /**
     * Returns the value of the specified character when using the "simple"
     * encoding from the Google Chart API.
     *
     * @param c  the character.
     *
     * @return The value.
     */
    private static int googleValue(char c) {
        if (c == '_') {
            return Integer.MAX_VALUE;
        }
        else if (c >= 'a' && c <= 'z') {
            return c - 97 + 26;
        }
        else if (c >= 'A' && c <= 'Z') {
            return c - 65;
        }
        else if (c >= '0' && c <= '9') {
            return c - 48 + 52;
        }
        else {
            throw new IllegalArgumentException("Invalid character '"
                    + c + "'.");
        }
    }

    /**
     * Returns the value of the specified character when using the "extended"
     * encoding from the Google Chart API.
     *
     * @param c  the character.
     *
     * @return The value.
     */
    private static int googleValueExt(char c) {
        if (c == '_') {
            return Integer.MAX_VALUE;
        }
        else if (c >= 'a' && c <= 'z') {
            return c - 97 + 26;
        }
        else if (c >= 'A' && c <= 'Z') {
            return c - 65;
        }
        else if (c >= '0' && c <= '9') {
            return c - 48 + 52;
        }
        else if (c == '-') {
            return 62;
        }
        else if (c == '.') {
            return 63;
        }
        else {
            throw new IllegalArgumentException("Invalid character '"
                    + c + "'.");
        }
    }

    /**
     * Parses a data scaling string. The returned array will always have a size
     * multiple of 2. If the passed in string is <code>null</code> or empty this
     * method returns {0, 100}. The values at odd positions are minimum values
     * while the values at even positions are the max values.
     *
     * @param scalingStr the scaling string.
     * @return the array of scaling values.
     */
    public static float[] parseDataScaling(String scalingStr) {
    	if (scalingStr == null) {
    		return new float[] {0f, 100.0f};
    	}
    	String[] parts = scalingStr.split(",");
    	if (parts.length == 0) {
    		return new float[] {0f, 100.0f};
    	}
    	// Make result array length at least as long as parts and multiple of 2
    	float[] result = new float[((parts.length + 1) >> 1) << 1];
    	for (int i = 0; i < parts.length; i++) {
    		if ("".equals(parts[i])) {
    			result[i] = 0f;
    		}
    		else {
    			result[i] = Float.parseFloat(parts[i]);
    		}
    	}
    	if (result.length > parts.length) {
    		result[result.length - 1] = 100.0f;
    	}
		return result;
	}

    private static float getScaling(float[] scaling, int datasetIndex,
    		int offset) {
    	int i = datasetIndex * 2;
    	if (i < scaling.length) {
    		return scaling[i + offset];
    	}
    	return scaling[scaling.length - 2 + offset];
    }

    private static float getScalingMin(float[] scaling, int datasetIndex) {
    	return getScaling(scaling, datasetIndex, 0);
    }

    private static float getScalingMax(float[] scaling, int datasetIndex) {
    	return getScaling(scaling, datasetIndex, 1);
    }

    /**
     * Parses a dataset from the incoming string (which may be encoded using
     * the 'simple', 'extended' or 'text' encodings from the Google Chart API).
     *
     * @param dataStr  the data string (<code>null</code> not permitted).
     * @param scalingStr  the scaling.
     *
     * @return A pie dataset.
     */
    public static PieDataset parsePieDataset(String dataStr,
    		String scalingStr) {
        if (dataStr == null) {
            throw new IllegalArgumentException("Null 'dataStr' argument.");
        }
        DefaultPieDataset result = new DefaultPieDataset();
        List dataList = null;
        if (dataStr.startsWith("s:")) { // simple encoding
            dataList = parseSimpleData(dataStr);
        }
        else if (dataStr.startsWith("t:")) {
        	float[] scaling = parseDataScaling(scalingStr);
            dataList = parseTextData(dataStr, scaling);
        }
        else if (dataStr.startsWith("e:")) {
            dataList = parseExtendedData(dataStr);
        }
        else {
            throw new RuntimeException("Unrecognised data encoding: "
                    + dataStr.substring(0, 2));
        }

        if (dataList.size() > 0) {
            List series = (List) dataList.get(0);
            int s = 1;
            Iterator iterator = series.iterator();
            while (iterator.hasNext()) {
                Number i = (Number) iterator.next();
                result.setValue("S" + s, i);
                s++;
            }
        }
        else {
            throw new RuntimeException("Empty data list.");
        }
        return result;
    }

    /**
     * Parses a category dataset for use in bar charts.
     *
     * @param dataStr  the data string (<code>null</code> not permitted).
     * @param scalingStr  the scaling string.
     *
     * @return A category dataset.
     */
    public static DefaultCategoryDataset parseCategoryDataset(String dataStr,
    		String scalingStr) {
        if (dataStr == null) {
            throw new IllegalArgumentException("Null 'dataStr' argument.");
        }
        List dataList = null;
        if (dataStr.startsWith("s:")) { // simple encoding
            dataList = DataUtilities.parseSimpleData(dataStr);
        }
        else if (dataStr.startsWith("t:")) {
        	float[] scaling = parseDataScaling(scalingStr);
            dataList = DataUtilities.parseTextData(dataStr, scaling);
        }
        else if (dataStr.startsWith("e:")) {
            dataList = DataUtilities.parseExtendedData(dataStr);
        }
        else {
            throw new RuntimeException("Unrecognised data encoding: "
                    + dataStr.substring(0, 2));
        }

        DefaultCategoryDataset result = new DefaultCategoryDataset();
        for (int i = 0; i < dataList.size(); i++) {
            List values = (List) dataList.get(i);
            for (int j = 0; j < values.size(); j++) {
                Number n = (Number) values.get(j);
                if (n != null) {
                    double y = n.doubleValue();
                    result.setValue(y, String.valueOf(i), String.valueOf(j));
                }
                else {
                    result.setValue(n, String.valueOf(i), String.valueOf(j));
                }
            }
        }

        return result;
    }

    /**
     * Parses a dataset from the incoming string.  The data encoding only
     * specifies the y-values.
     *
     * @param dataStr  the data string (<code>null</code> not permitted).
     * @param scalingStr  the scaling string.
     *
     * @return A dataset.
     */
    public static XYDataset parseXYDataset(String dataStr, String scalingStr) {

        if (dataStr == null) {
            throw new IllegalArgumentException("Null 'dataStr' argument.");
        }
        List dataList;
        if (dataStr.startsWith("s:")) { // simple encoding
            dataList = DataUtilities.parseSimpleData(dataStr);
        }
        else if (dataStr.startsWith("t:")) {
        	float[] scaling = parseDataScaling(scalingStr);
            dataList = DataUtilities.parseTextData(dataStr, scaling);
        }
        else if (dataStr.startsWith("e:")) {
            dataList = DataUtilities.parseExtendedData(dataStr);
        }
        else {
            throw new RuntimeException("Unrecognised data encoding: "
                    + dataStr.substring(0, 2));
        }
        XYSeriesCollection result = new XYSeriesCollection();
        for (int s = 0; s < dataList.size(); s++) {
            List series = (List) dataList.get(s);
            XYSeries ss = new XYSeries("Series " + (s + 1));
            int itemCount = series.size();
            double increment = 1.0 / (Math.max(itemCount - 1, 1));
            for (int i = 0; i < itemCount; i++) {
                Number n = (Number) series.get(i);
                if (n != null) {
                    double y = n.doubleValue();
                    ss.add(i * increment, y);
                }
                else {
                    ss.add(i * increment, null);
                }
            }
            result.addSeries(ss);
        }
        return result;
    }

    /**
     * Parses a dataset from the incoming string.  The data encoding specifies
     * both the x-values and the y-values, in consecutive series.
     *
     * @param dataStr  the data string (<code>null</code> not permitted).
     * @param scalingStr  the scaling string.
     *
     * @return A dataset.
     */
    public static XYDataset parseXYDataset2(String dataStr, String scalingStr) {

        if (dataStr == null) {
            throw new IllegalArgumentException("Null 'dataStr' argument.");
        }
        List dataList;
        if (dataStr.startsWith("s:")) { // simple encoding
            dataList = DataUtilities.parseSimpleData(dataStr);
        }
        else if (dataStr.startsWith("t:")) {
        	float[] scaling = parseDataScaling(scalingStr);
            dataList = DataUtilities.parseTextData(dataStr, scaling);
        }
        else if (dataStr.startsWith("e:")) {
            dataList = DataUtilities.parseExtendedData(dataStr);
        }
        else {
            throw new RuntimeException("Unrecognised data encoding: "
                    + dataStr.substring(0, 2));
        }
        XYSeriesCollection result = new XYSeriesCollection();
        for (int s = 0; s < dataList.size(); s = s + 2) {
            List xSeries = (List) dataList.get(s);
            List ySeries = (List) dataList.get(s + 1);
            XYSeries ss = new XYSeries("Series " + (s / 2));
            int itemCount = xSeries.size();
            // check for special case of x-values = -1
            boolean implicitXValues = false;
            if (itemCount == 1) {
                itemCount = ySeries.size();
                implicitXValues = true;
            }
            for (int i = 0; i < itemCount; i++) {
                double xx;
                if (implicitXValues) {
                    xx = (i + 0.0) / Math.max(itemCount - 1, 1);
                }
                else {
                    Number x = (Number) xSeries.get(i);
                    xx = x.doubleValue();
                }
                Number y = (Number) ySeries.get(i);
                if (y != null) {
                    ss.add(xx, y.doubleValue());
                }
                else {
                    ss.add(xx, null);
                }
            }
            result.addSeries(ss);
        }
        return result;
    }

    /**
     * Parses a dataset from the incoming string.  The data encoding specifies
     * both the x-values and the y-values, in consecutive series, with an
     * optional third series supplying the relative marker sizes.
     *
     * @param dataStr  the data string.
     * @param scalingStr  the scaling string.
     *
     * @return A dataset.
     */
    public static XYSeriesCollection parseScatterDataset(String dataStr,
    		String scalingStr) {

        if (dataStr == null) {
            throw new IllegalArgumentException("Null 'dataStr' argument.");
        }
        List dataList;
        if (dataStr.startsWith("s:")) { // simple encoding
            dataList = DataUtilities.parseSimpleData(dataStr);
        }
        else if (dataStr.startsWith("t:")) {
        	float[] scaling = parseDataScaling(scalingStr);
            dataList = DataUtilities.parseTextData(dataStr, scaling);
        }
        else if (dataStr.startsWith("e:")) {
            dataList = DataUtilities.parseExtendedData(dataStr);
        }
        else {
            throw new RuntimeException("Unrecognised data encoding: "
                    + dataStr.substring(0, 2));
        }
        XYSeriesCollection result = new XYSeriesCollection();
        List xSeries = (List) dataList.get(0);
        List ySeries = (List) dataList.get(1);
        XYSeries ss = new XYSeries("Series");
        int itemCount = xSeries.size();

        for (int i = 0; i < itemCount; i++) {
            Number x = (Number) xSeries.get(i);
            double xx = x.doubleValue();
            Number y = (Number) ySeries.get(i);
            if (y != null) {
                double yy = y.doubleValue();
                ss.add(xx, yy);
            }
            else {
                ss.add(xx, null);
            }
            result.addSeries(ss);
        }

        // is there a third series?
        List zSeries = null;
        if (dataList.size() > 2) {
            zSeries = (List) dataList.get(2);
            XYSeries sz = new XYSeries("Z");
            for (int i = 0; i < zSeries.size(); i++) {
                sz.add(i, (Number) zSeries.get(i));
            }
            // FIXME: we're not doing anything with this
        }

        return result;
    }

}
